airbyte-cdk/python/airbyte_cdk/sources/connector_state_manager.py

```diff
@@ -14,41 +27,117 @@ class ConnectorStateManager:
     interface. It also provides methods to extract and update state
     """
 
-    # In the immediate, we only persist legacy which will be used during abstract_source.read(). In the subsequent PRs we will
-    # initialize the ConnectorStateManager according to the new per-stream interface received from the platform
-    def __init__(self, state: Union[List[AirbyteStateMessage], MutableMapping[str, Any]] = None):
-        if not state:
-            self.legacy = {}
-        elif self.is_migrated_legacy_state(state):
-            # The legacy state format received from the platform is parsed and stored as a single AirbyteStateMessage when reading
-            # the file. This is used for input backwards compatibility.
-            self.legacy = state[0].data
-        elif isinstance(state, MutableMapping):
-            # In the event that legacy state comes in as its original JSON object format, no changes to the input need to be made
-            self.legacy = state
-        else:
+    def __init__(self, stream_instance_map: Mapping[str, Stream], state: Union[List[AirbyteStateMessage], MutableMapping[str, Any]] = None):
+        shared_state, per_stream_states = self._extract_from_state_message(state, stream_instance_map)
+
+        # We explicitly throw an error if we receive a GLOBAL state message that contains a shared_state because API sources are
+        # designed to checkpoint state independently of one another. API sources should never be emitting a state message where
+        # shared_state is populated. Rather than define how to handle shared_state without a clear use case, we're opting to throw an
+        # error instead and if/when we find one, we will then implement processing of the shared_state value.
+        if shared_state:
+            raise ValueError("API source connectors do not currently support GLOBAL AirbyteStateMessages that contain a shared_state")
```

Comment:
```suggestion
        per_stream_states = self._extract_from_state_message(state, stream_instance_map)
```
Could we throw the error from `_extract_from_state_message` and only return the per stream states?

Reply:
The idea here is to actually fail loudly because `shared_states` is unexpected output and we shouldn't try to continue the sync if we received this kind of input. I'm also on the side of leaving the error validation/message in the constructor since it allows the `_extract_from_state_message` to be solely responsible for just retrieving necessary info, and the constructor can be in charge of validating what it gets back